index.js

```diff
@@ -18,149 +10,197 @@ var mkdirp = require('mkdirp');
- * @param {object} `options` Options to pass to [fs.writeFile][fs]{#fs_fs_writefile_file_data_options_callback} and/or [mkdirp][]
+ * @name write
+ * @param {String} `filepath` file path.
+ * @param {String|Buffer|Uint8Array} `data` Data to write.
+ * @param {Object} `options` Options to pass to [fs.writeFile][writefile]
  * @param {Function} `callback` (optional) If no callback is provided, a promise is returned.
  * @api public
  */
 
-function writeFile(filepath, data, options, cb) {
+const write = (filepath, data, options, callback) => {
   if (typeof options === 'function') {
-    cb = options;
+    callback = options;
     options = {};
   }
 
-  if (typeof cb !== 'function') {
-    return writeFile.promise.apply(null, arguments);
-  }
+  const opts = { encoding: 'utf8', ...options };
+  const destpath = opts.increment ? incrementName(filepath) : filepath;
+  const result = { path: destpath, data };
 
-  if (typeof filepath !== 'string') {
-    cb(new TypeError('expected filepath to be a string'));
-    return;
+  if (opts.overwrite === false && exists(filepath, destpath)) {
+    throw new Error('File already exists: ' + destpath);
   }
 
-  mkdirp(path.dirname(filepath), options, function(err) {
-    if (err) {
-      cb(err);
-      return;
-    }
+  const promise = mkdir(path.dirname(destpath), { recursive: true, ...options })
+    .then(() => {
+      return new Promise((resolve, reject) => {
+        fs.createWriteStream(destpath, opts)
+          .on('error', err => reject(err))
+          .on('close', resolve)
+          .end(ensureNewline(data, opts));
+      });
+    });
 
-    var preparedData = data;
-    if (options && options.ensureNewLine && data.slice(-1) !== "\n") {
-      preparedData += "\n";
-    }
+  if (typeof callback === 'function') {
+    promise.then(() => callback(null, result)).catch(callback);
+    return;
+  }
 
-    fs.writeFile(filepath, preparedData, options, cb);
-  });
+  return promise.then(() => result);
 };
 
 /**
- * The promise version of [writeFile](#writefile). Returns a promise.
+ * The synchronous version of [write](#write). Returns undefined.
  *
  * ```js
- * var writeFile = require('write');
- * writeFile.promise('foo.txt', 'This is content...')
- *   .then(function() {
- *     // do stuff
- *   });
+ * const write = require('write');
+ * write.sync('foo.txt', 'This is content...');
  * ```
- * @name .promise
- * @param {string|Buffer|integer} `filepath` filepath or file descriptor.
- * @param {string|Buffer|Uint8Array} `val` String or buffer to write to disk.
- * @param {object} `options` Options to pass to [fs.writeFile][fs]{#fs_fs_writefile_file_data_options_callback} and/or [mkdirp][]
- * @return {Promise}
+ * @name .sync
+ * @param {String} `filepath` file path.
+ * @param {String|Buffer|Uint8Array} `data` Data to write.
+ * @param {Object} `options` Options to pass to [fs.writeFileSync][writefilesync]
+ * @return {undefined}
```

Comment:
The return value is an object with `path` and `data`.

Reply:
done